implementation-guide/Implementation Guide.md

```diff
@@ -0,0 +1,632 @@
+```
+
+The message looks good, it seems to have made a new resolver object and exported that at the next available position and put it in the session's export table. After some time a new message comes in and it reads:
+
+```
+<op:deliver-only    <desc:export 2>
+                    ['fulfill <desc:import-promise 2>]>
+```
+
+Cool! but wait, `import-promise`? The robot has created a promise to beep and send that. Okay, that's fine, she just has to listen to it like before. She looks at the CapTP specification and sees the `op:listen` operation, this lets her give a promise and setup a listener so it can respond. The `op:listen` operation looks like this:
+
+```
+<op:listen to-desc           ; desc:export | desc:answer
+           listen-desc       ; desc:import-object
+           wants-partial?>   ; boolean
+```
+
+So Alisha needs to send this operation to the promise which was exported to her with an object that wants to be notified. Fortunately that's just the same notification object that promises she implemented before has. Finally the promise has this `wants-partial?` boolean which will decide if she'll get updates even if it's just to another promise. She decides at her CapTP boundary look for `desc:import-promise` references and create a promise pair, she'll replace the references to the remote promise with her local one and transmit an `op:listen` with the resolver to the vow in the `listen-desc`, for `wants-partial?` she'll just leave that at false for now.
+
+She sets up a promise pair and transmits `op:listen` for the promise she got earlier for the beep, the message like this:
+
+```
+<op:listen  <desc:export 2>
+            <desc:import-object 3>
+            true>
+```
+
+After another while she gets another message with the resolution to that promise:
+```
+<op:deliver-only    <desc:import-object 3>
+                    ['fulfill "beeeep!"]>
+```
+
+Oh cool! that all worked. Now she has promises though she needs to modify how her `op:abort` works, before it just terminated the session if she got one, but now she has promises she needs to ensure they break if the session ends. She sets up the promise creation code to add her promise to a set of unresolved promises that her `op:abort` handling code can access. When the resolver gets the resolution, it removes itself from the set of unresolved promises. She next makes it so that when an `op:abort` comes in, she iterates through the set sending the resolver objects a break with a network partition error. This should ensure promises are handled correctly.
+
+Promises might not only be fulfilled through, lets imagine that Alisha sents another message to move forward. Lets assume the robot can't do that, it doesn't really matter why, maybe the robot physically isn't able to move forward for there being an obstical, maybe it's got low power, or some other reason entirely, but the promise she gets back instead of being resolved as fulfilled instead gets broken:
```

Comment:
```suggestion
Promises might not only be fulfilled through, lets imagine that Alisha sents another message to move forward. Lets assume the robot can't do that, it doesn't really matter why, maybe the robot physically isn't able to move forward for there being an obstacle, maybe it's got low power, or some other reason entirely, but the promise she gets back instead of being resolved as fulfilled instead gets broken:
```